File: ctci/chapter_1_arrays_strings/js/1.2-checkPermutation.js
/* O(nlog(n)) solution due to initial sort */

/**
 * Sorts string.
 * @param   {string}  str
 * @returns {string}
 */
const sorted = str =>
  str
    .split("")
    .sort()
    .join("")

/**
 * Determines if two strings are permutations of each other.
 * @param   {string}  str1
 * @param   {string}  str2
 * @returns {boolean}
 */
const checkPermutationSort = (str1, str2) => {
  if (str1.length != str2.length) return false
  return sorted(str1) === sorted(str2)
}

/* O(n) solution */

/**
 * Creates hash table of string.
 * @param   {string}  str
 * @returns {object}
 */
const createMap = str => {
  return str.split("").reduce((a, b) => {
    a[b] ? a[b]++ : (a[b] = 1)
    return a
  }, {})
}

/**
 * Determines if two strings are permutations of each other.
 * @param   {string}  str1
 * @param   {string}  str2
 */
const checkPermutation = (str1, str2) => {
  if (str1.length != str2.length) return false
  const map = createMap(str1)
  for (let ch of str2) {
    if (!map[ch]) return false
    map[ch]--
    /* If array map has a negative value, there is
     * a mismatch in characters. */
    if (map[ch] < 0) return false
  }
  return true
}

console.log(checkPermutation("dog", "god")) // true
console.log(checkPermutation("dog", "God")) // false
console.log(checkPermutation("dog", "good")) // false
console.log(checkPermutation("mmmmm", "mmmmm")) // true
console.log(checkPermutation("mmmmm", "mmmmn")) // false

File: ctci/chapter_1_arrays_strings/python/1.2-check_permutation.py
# O(nlog(n)) solution due to initial sort


def check_permutation_sort(str1: str, str2: str) -> bool:
    """
    Determines if two strings are permutations of each other.
    """
    if len(str1) != len(str2):
        return False
    return ''.join(sorted(str1)) == ''.join(sorted(str2))


# O(n) solution


def create_map(string: str) -> dict:
    """
    Creates hash table of string.
    """
    obj = {}
    for ch in string:
        if ch in obj:
            obj[ch] += 1
        else:
            obj[ch] = 1
    return obj


def check_permutation(str1: str, str2: str) -> bool:
    """
    Determines if two strings are permutations of each other.
    """
    if len(str1) != len(str2):
        return False
    str_map = create_map(str1)
    for ch in str2:
        if ch not in str_map:
            return False
        str_map[ch] -= 1
        # if array map has a negative value, there is a mismatch in
        # characters.
        if str_map[ch] < 0:
            return False
    return True


if __name__ == 'main':
    print(check_permutation('dog', 'god'))  # true
    print(check_permutation('dog', 'God'))  # false
    print(check_permutation('dog', 'good'))  # false
    print(check_permutation('mmmmm', 'mmmmm'))  # true
    print(check_permutation('mmmmm', 'mmmmn'))  # true
